rust/protocol/src/curve.rs

```diff
@@ -1,56 +1,98 @@
+// Copyright 2020-2022 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
-pub(crate) mod curve25519;
-
+//! Cryptographic primitives for asymmetric keys.
+//!
+//! Some example operations:
+//!```
+//! use libsignal_protocol::KeyPair;
+//! use std::collections::HashSet;
+//!
+//! let alice = KeyPair::generate(&mut rand::thread_rng());
+//! assert!(alice == alice.clone());
+//! let bob = KeyPair::generate(&mut rand::thread_rng());
+//! assert!(alice != bob);
+//!
+//! // Keys can be hashed and put in sets.
+//! let key_set: HashSet<KeyPair> = [alice, bob].iter().cloned().collect();
+//! assert!(key_set.contains(&alice));
+//! assert!(key_set.contains(&bob));
+//!```
+
+#![warn(missing_docs)]
+
+pub mod curve25519;
+
+use crate::utils::constant_time_cmp;
 use crate::{Result, SignalProtocolError};
 
 use std::cmp::Ordering;
-use std::convert::TryFrom;
+use std::convert::{TryFrom, TryInto};
 use std::fmt;
+use std::hash;
 
 use arrayref::array_ref;
+use displaydoc::Display;
+use num_enum::{IntoPrimitive, TryFromPrimitive, TryFromPrimitiveError};
 use rand::{CryptoRng, Rng};
 use subtle::ConstantTimeEq;
 
-#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+/// Map the variant of key being used to and from a [`u8`].
+///
+///```
+/// # fn main() -> libsignal_protocol::error::Result<()> {
+/// use libsignal_protocol::KeyType;
+/// use std::convert::{TryFrom, TryInto};
+///
+/// let encoded_key_type: u8 = KeyType::Curve25519.into();
+/// assert!(encoded_key_type == 0x05);
+///
+/// let original_key_type: KeyType = encoded_key_type.try_into()?;
+/// assert!(original_key_type == KeyType::Curve25519);
+///
+/// let bad_encoded_key: u8 = 0x27;
+/// assert!(KeyType::try_from(bad_encoded_key).is_err());
+/// # Ok(())
+/// # }
+///```
+#[derive(
+    Debug, Display, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, TryFromPrimitive, IntoPrimitive,
+)]
+#[ignore_extra_doc_attributes]
+#[repr(u8)]
 pub enum KeyType {
-    Djb,
-}
-
-impl fmt::Display for KeyType {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Debug::fmt(self, f)
-    }
+    /// <Curve25519 type key>
+    ///
+    /// See [Curve25519].
+    ///
+    /// [Curve25519]: https://en.wikipedia.org/wiki/Curve25519.
+    Curve25519 = 0x05,
 }
 
-impl KeyType {
-    fn value(&self) -> u8 {
-        match &self {
-            KeyType::Djb => 0x05u8,
-        }
+impl From<TryFromPrimitiveError<KeyType>> for SignalProtocolError {
+    fn from(err: TryFromPrimitiveError<KeyType>) -> Self {
+        SignalProtocolError::BadKeyType(err.number)
     }
 }
 
-impl TryFrom<u8> for KeyType {
-    type Error = SignalProtocolError;
-
-    fn try_from(x: u8) -> Result<Self> {
-        match x {
-            0x05u8 => Ok(KeyType::Djb),
-            t => Err(SignalProtocolError::BadKeyType(t)),
-        }
-    }
+/// Interface for structs that perform operations parameterized by values of [`KeyType`].
+pub trait Keyed {
```

Comment:
Hm, this isn't really the right name for the trait. Public keys and private keys and key pairs aren't "keyed", they're…"typed"?

That said, though, I think this is a reasonable trait, but until we have a use case for it I'd rather not add it, and that won't happen until we support another key type. It might turn out we need different operations besides just `key_type`.